### ISSUE TITLE
Parallel code coverage

### ISSUE THREAD

**Issue by Jean85** [facile-it/paraunit#11]
I've started to work to a new feature, a command that produces code coverage with Paraunit.

I'll merge the branch ASAP, and I'll use this issue to track the progress.
- [x] Big refactoring
- [x] First working version
- [x] Support for PHPDBG (PHP7)
- [x] More testing!
- [x] Merge `feature-coverage` to master
- [x] Create 0.7 milestone and add this issue to it
- [x] Check for syntax before merging the coverage results
- [ ] Try to have parallel result merging (recursive + binary approach?)
- [ ] Check for performance issues

[...]
- [ ] Tag as 0.7 and close this issue!

**Comment by fntlnz** [comment]
WOW! Very interesting!

**Comment by Jean85** [comment]
We are investigating, because the performance of this feature seems to stay poor. It always stays with just one core occupied, and never do stuff in parallel, excepts some rare peaks.

I first tought that xDebug was forcing the inline execution because it was waiting for some resource.
Working with @fntlnz and @taueres , we fear that the resource in question is the .php file(s). If xDebug puts a read lock on the source code to do the coverage analysis, it's a big setback for this.

I want to understand if this is the real issue, and then decide if we need to postpone o cancel this feature.

**Comment by ranpafin** [comment]
We could check with lslocks if those files are really locked and by what process and go from there.

**Comment by Jean85** [comment]
I've tried to look at `htop` during the execution of the `coverage` command, and there were always at least 3 process running, so XDebug it's not locking anything!

I want to try it under PHP7, because it uses PHPDBG for the coverage. That will need a new strategy for the process launch, because the command line needs to be heavily altered..

**Comment by fntlnz** [comment]
It currently uses phpdbg just because xdebug isn't yet available on php7. However I think that having both can be a fair thing.

**Comment by Jean85** [comment]
Yeah sure!! I'm thinking about a strategy pattern or some DI magic to have both. But first, we need to know if either tool is viable with Paraunit...
As of now, XDebug doesn't seem usable for me, it's even slower than vanilla PHPUnit (at least with unit tests, I need to try it with functional ones); if we don't identify (and resolve) any bottleneck, I think we shouldn't implement it.

I will try PHPDBG further, if it's faster (or it hasn't any unknown bottleneks), it could be our chosen solution.

**Comment by ranpafin** [comment]
I've tried running the coverage with one of the big project (SF2 suite) and the cpu it's always working at full load.

In this dstat report you can see the the idle (idl) cpu time is 0 which means that all the cpu's always have something to do.
If we have performance issue for the parallel coverage is not in the paralization itself, nor is a locking issue else there would be some idle cpu cycle waiting for locks to be released.

```
----total-cpu-usage---- -dsk/total- -net/total- ---paging-- ---system--
usr sys idl wai hiq siq| read  writ| recv  send|  in   out | int   csw 
 97   3   0   0   0   0|   0   164k|5014B  602B|   0     0 |2895  5083 
 97   3   0   0   0   0|   0  2760k|2879B 3969B|   0     0 |2813  4784 
 98   2   0   0   0   0|   0     0 |  15k 1234B|   0     0 |2979  5057 
 98   2   0   0   0   0|   0   152k|2294B  198B|   0     0 |2862  4805 
 96   4   0   0   0   0|   0   172k|4436B 1881B|   0     0 |3130  5499 
 93   7   0   0   0   0|   0  1056k|2942B 1552B|   0     0 |3057  5793 
 96   4   0   0   0   0|   0   528k|5257B 2718B|   0     0 |2730  4843 
 94   6   0   0   0   0|4096B  152k|2479B  198B|   0     0 |3060  4997 
 94   5   0   0   0   0|   0     0 |4363B  264B|   0     0 |2915  4889 
 91   8   0   0   0   0|   0     0 |2412B  218B|   0     0 |2977  5302 
```

But then i went trough the much more faster-to-execute unit tests an there the situation is a lot different:

```
usr sys idl wai hiq siq| read  writ| recv  send|  in   out | int   csw 
 50  22  29   0   0   0|   0     0 |1671B  215B|   0     0 |4271    32k
 52  19  28   0   0   1|   0   332k|3800B  408B|   0     0 |3991    20k
 44  22  34   0   0   0|   0   204k|1560B  198B|   0     0 |4385    22k
 46  19  35   0   0   1|   0    72k|2465B  198B|   0     0 |4084    25k
 60  18  21   0   0   0|  16k 2600k|  78k   35k|   0     0 |3947    21k
 45  21  34   0   0   0|  12k  396k| 132k   22k|   0     0 |4397    23k
 38  23  39   0   0   1|4096B    0 |3443B  690B|   0     0 |4567    34k
 36  26  39   0   0   0|   0  8192B|2064B  264B|   0     0 |4237    41k
 37  26  37   0   0   0|   0   112k|2243B  198B|   0     0 |4212    37k
```

As you can see the csw are skyrocketing which means that we generate a huge number of processes that do only little things so some (many) CPU cycle are wasted in moving in and out the stuff.

Also watching the report for a couple of minutes (here it's not clear) i could see that when only when a test-file is completed something i writed down to disk while in the second case the disk i costantly used. So more idle time.

i will further investigate this... But it's not good news so fare, we may need to change the logic or insert some kind of buffer.

**Comment by Jean85** [comment]
Seems to me that the too-many-processes problem for the unit tests is still present here, as in the simple run command.
Maybe a worker approach could solve this?

**Comment by ranpafin** [comment]
Could be an idea... It would take some time to come up with a prototype, do you think we should opt-out this feature from  0.5?

**Comment by Jean85** [comment]
Yeah, maybe... I was thinking the same thing!
Also, if the situation for coverage performance and usability will be a lot different between PHP 5 and 7, we could consider to implement this feature only for PHP7.

**Comment by Jean85** [comment]
Removing this from 0.5. I will continue reporting our findings here anyway.

**Comment by Jean85** [comment]
I've started a new branch from master https://github.com/facile-it/paraunit/tree/feature-coverage-new-approach with commit 4c42152

I'm trying a new approach, that's more aligned to the recend developments on master. I will try to make a first working version on PHP5, and enable PHPDbg support later for PHP7; I'm hoping on a flexible solution to support both, since only one of the is supported on each major of PHP, due to driver limitation on the PHPUnit side.

If this looks promising, I will open a WIP PR, this issue is too messed up and confused right now.

**Comment by Jean85** [comment]
Today I've tackled the issue once more. The branch has been rebased and looks promising, but I think that I need a last refactor to accomplish it.

Basically, I can't have the container compiled in the same way for boh commands (run and coverage). I need to pospone the container build inside the commands, and set the CLI arguments/options as container parameters.

In this way, I can build two different (but very  similar) containers, where the only differences are the non-common services. Also, I can remove listeners instead of setting state to them to enable\disable them.

**Comment by Jean85** [comment]
I've launched the coverage over Paraunit itself and it went perfectly!

I've also launched it on a big private project, and it crashed. The issue was a truncated coverage result file, due probably to a memory error or some other abnormal termination.

I need to protect the coverage result merger from this kind of problems. Excluding the coverage result of a malformed file should not be an issue.

**Comment by Jean85** [comment]
I've fully rebased this to the `feature-coverage-new-approach-rebased` branch. Now I need to add the last touches like the syntax check, that can be done with a simple `php -l` process. I hope to complete this ASAP!

**Comment by ranpafin** [comment]
@Jean85 Great! Can't wait to try this on our company test-suite!

**Comment by Jean85** [comment]
This feature is now on master, after merging #59!!! 

YAY!!!! :tada:

**Comment by Jean85** [comment]
I've tested this feature on a small project and it works flawlessly, identical HTML generated (apart from the signature that reports the coverage driver used).

I'm still having issues on a bigger project, but I think that is just a smaller issue. 579f9d0 should help, also I will tackle #69 and #70
